internal/server/data/passwordresettoken.go

```diff
@@ -4,63 +4,86 @@ import (
 	"errors"
 	"time"
 
-	"gorm.io/gorm"
-
 	"github.com/infrahq/infra/internal"
 	"github.com/infrahq/infra/internal/generate"
 	"github.com/infrahq/infra/internal/logging"
 	"github.com/infrahq/infra/internal/server/models"
 	"github.com/infrahq/infra/uid"
 )
 
-func CreatePasswordResetToken(db GormTxn, user *models.Identity, ttl time.Duration) (*models.PasswordResetToken, error) {
+type passwordResetToken struct {
+	ID uid.ID
+	models.OrganizationMember
+
+	Token      string
+	IdentityID uid.ID
+	ExpiresAt  time.Time
+}
+
+func (passwordResetToken) Table() string {
+	return "password_reset_tokens"
+}
+
+func (p passwordResetToken) Columns() []string {
+	return []string{"expires_at", "id", "identity_id", "organization_id", "token"}
+}
+
+func (p passwordResetToken) Values() []any {
+	return []any{p.ExpiresAt, p.ID, p.IdentityID, p.OrganizationID, p.Token}
+}
+
+func (p *passwordResetToken) ScanFields() []any {
+	return []any{&p.ExpiresAt, &p.ID, &p.IdentityID, &p.OrganizationID, &p.Token}
+}
+
+func (p *passwordResetToken) OnInsert() error {
+	return nil
+}
+
+func CreatePasswordResetToken(db GormTxn, user *models.Identity, ttl time.Duration) (string, error) {
 	tries := 0
+	var ucErr UniqueConstraintError
+
 retry:
 	token, err := generate.CryptoRandom(10, generate.CharsetAlphaNumeric)
 	if err != nil {
-		return nil, err
+		return "", err
 	}
 
-	prt := &models.PasswordResetToken{
+	prt := &passwordResetToken{
 		ID:         uid.New(),
 		Token:      token,
 		IdentityID: user.ID,
 		ExpiresAt:  time.Now().Add(ttl).UTC(),
 	}
 
 	tries++
-	if err = save(db, prt); err != nil {
-		// TODO: must use errors.As for error types
-		if tries <= 3 && errors.Is(err, UniqueConstraintError{}) {
```

Comment:
I also resolved this TODO at the same time. Previously the `errors.Is` would never return true because the `err` would have the `Table` and `Column` field set.